File: win/system/backend/app/core/security.py
"""
Security utilities for authentication and authorization
Handles JWT tokens, password hashing, and Turnstile verification
"""

import httpx
import logging
from datetime import datetime, timedelta
from typing import Optional
from jose import jwt, JWTError
from passlib.hash import bcrypt
from fastapi import HTTPException, status

from app.core.config import (
    JWT_SECRET, 
    JWT_ALGORITHM, 
    ACCESS_TOKEN_EXPIRE_MINUTES,
    TURNSTILE_SECRET
)

logger = logging.getLogger(__name__)

# ============================================================================
# PASSWORD HASHING
# ============================================================================

def hash_password(password: str) -> str:
    """Hash a password using bcrypt"""
    return bcrypt.hash(password)

def verify_password(plain_password: str, hashed_password: str) -> bool:
    """Verify a password against a hash"""
    return bcrypt.verify(plain_password, hashed_password)

# ============================================================================
# JWT TOKEN MANAGEMENT
# ============================================================================

def create_access_token(data: dict) -> str:
    """Create a JWT access token"""
    to_encode = data.copy()
    expire = datetime.utcnow() + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    to_encode.update({"exp": expire})
    
    encoded_jwt = jwt.encode(to_encode, JWT_SECRET, algorithm=JWT_ALGORITHM)
    return encoded_jwt

def decode_access_token(token: str) -> dict:
    """Decode and validate a JWT token"""
    try:
        payload = jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])
        return payload
    except JWTError:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid or expired token",
            headers={"WWW-Authenticate": "Bearer"}
        )

# ============================================================================
# TURNSTILE VERIFICATION
# ============================================================================

async def verify_turnstile_token(token: str, remote_ip: Optional[str] = None) -> bool:
    """
    Verify Cloudflare Turnstile token
    
    Args:
        token: Turnstile response token
        remote_ip: Client IP address (optional)
    
    Returns:
        True if verification succeeds
    
    Raises:
        HTTPException: If verification fails or server error
    """
    if not TURNSTILE_SECRET:
        logger.error("DOUGH_TURNSILE_SECRET environment variable not set")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Server configuration error"
        )
    
    verify_url = "https://challenges.cloudflare.com/turnstile/v0/siteverify"
    
    data = {
        "secret": TURNSTILE_SECRET,
        "response": token,
    }
    
    if remote_ip:
        data["remoteip"] = remote_ip
    
    try:
        async with httpx.AsyncClient() as client:
            response = await client.post(verify_url, data=data)
            response.raise_for_status()
            
            result = response.json()
            
            if not result.get("success", False):
                logger.warning(f"Turnstile verification failed: {result.get('error-codes', [])}")
                return False
            
            logger.info("Turnstile verification successful")
            return True
            
    except httpx.RequestError as e:
        logger.error(f"Failed to verify Turnstile token: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to verify security token"
        )
    except Exception as e:
        logger.error(f"Unexpected error during Turnstile verification: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Security verification error"
        )

File: system/backend/src/core/security.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */

/**
 * Security utilities for authentication and authorization.
 * Handles JWT tokens, password hashing, and Turnstile verification.
 */

import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import axios from 'axios';

import { JWT_SECRET, JWT_ALGORITHM, ACCESS_TOKEN_EXPIRE_MINUTES, TURNSTILE_SECRET } from './config.js';
import { HttpError } from './errors.js';

const BCRYPT_ROUNDS = 10;

// ============================================================================
// PASSWORD HASHING
// ============================================================================

/**
 * Hash a password using bcrypt.
 *
 * bcrypt only considers the first 72 bytes of input. We truncate to 72
 * bytes to keep behaviour stable and avoid errors on long passwords,
 * mirroring the Python implementation.
 */
export async function hashPassword(password: string): Promise<string> {
  const truncated = Buffer.from(password, 'utf-8').subarray(0, 72).toString('utf-8');
  return bcrypt.hash(truncated, BCRYPT_ROUNDS);
}

/**
 * Verify a password against a hash.
 *
 * Returns false (rather than throwing) if the stored hash is missing or
 * malformed, e.g. a record left behind by a failed startup.
 */
export async function verifyPassword(plainPassword: string, hashedPassword?: string | null): Promise<boolean> {
  if (!hashedPassword) {
    return false;
  }
  try {
    const truncated = Buffer.from(plainPassword, 'utf-8').subarray(0, 72).toString('utf-8');
    return await bcrypt.compare(truncated, hashedPassword);
  } catch {
    return false;
  }
}

// ============================================================================
// JWT TOKEN MANAGEMENT
// ============================================================================

/** Create a JWT access token */
export function createAccessToken(data: Record<string, unknown>): string {
  return jwt.sign(data, JWT_SECRET, {
    algorithm: JWT_ALGORITHM,
    expiresIn: `${ACCESS_TOKEN_EXPIRE_MINUTES}m`,
  });
}

/** Decode and validate a JWT token */
export function decodeAccessToken(token: string): jwt.JwtPayload {
  try {
    const payload = jwt.verify(token, JWT_SECRET, { algorithms: [JWT_ALGORITHM] });
    if (typeof payload === 'string') {
      throw new Error('Unexpected string payload');
    }
    return payload;
  } catch {
    throw new HttpError(401, 'Invalid or expired token', { 'WWW-Authenticate': 'Bearer' });
  }
}

// ============================================================================
// TURNSTILE VERIFICATION
// ============================================================================

interface TurnstileResponse {
  success: boolean;
  'error-codes'?: string[];
}

/**
 * Verify a Cloudflare Turnstile token.
 *
 * @param token - Turnstile response token
 * @param remoteIp - Client IP address (optional)
 * @returns true if verification succeeds
 * @throws HttpError if verification fails or a server error occurs
 */
export async function verifyTurnstileToken(token: string, remoteIp?: string): Promise<boolean> {
  if (!TURNSTILE_SECRET) {
    console.error('TURNSILE_SECRET environment variable not set');
    throw new HttpError(500, 'Server configuration error');
  }

  const verifyUrl = 'https://challenges.cloudflare.com/turnstile/v0/siteverify';

  const params = new URLSearchParams({
    secret: TURNSTILE_SECRET,
    response: token,
  });

  if (remoteIp) {
    params.set('remoteip', remoteIp);
  }

  try {
    const { data } = await axios.post<TurnstileResponse>(verifyUrl, params);

    if (!data.success) {
      console.warn(`Turnstile verification failed: ${JSON.stringify(data['error-codes'] ?? [])}`);
      return false;
    }

    console.info('Turnstile verification successful');
    return true;
  } catch (err) {
    if (err instanceof HttpError) {
      throw err;
    }
    console.error(`Failed to verify Turnstile token: ${String(err)}`);
    throw new HttpError(500, 'Failed to verify security token');
  }
}